pgmpy/identification/instrument.py

```diff
@@ -0,0 +1,148 @@
+from networkx.algorithms.dag import descendants
+
+from pgmpy.base import DAG
+from pgmpy.global_vars import logger
+from pgmpy.identification import BaseIdentification
+from pgmpy.inference.CausalInference import CausalInference
+
+
+class InstrumentalVariables(BaseIdentification):
+
+    def __init__(self, variant=None, scaling_indicators=None):
+        self.supported_graph_types = DAG
+        self.variant = variant
+        self.scaling_indicators = scaling_indicators
+
+    def _get_scaling_indicators(self, causal_graph):
+        latent_variables = causal_graph.get_role("latents")
+        observed_nodes = causal_graph.get_role("observed")
+        scaling_indicators = {}
+
+        if self.scaling_indicators is not None:
+            return self.scaling_indicators
+
+        for node in latent_variables:
+            for neighbour in causal_graph.neighbors(node):
+                if neighbour in observed_nodes:
+                    scaling_indicators[node] = neighbour
+                    break
+
+        return scaling_indicators
+
+    def _iv_transformations(self, X, Y, causal_graph, scaling_indicators=None):
+        full_graph_ = causal_graph.copy()
+
+        exposures = full_graph_.get_role("exposures")
+        latent_variables = full_graph_.get_role("latents")
+        observed = full_graph_.get_role("observed")
+        scaling_indicators = self._get_scaling_indicators(causal_graph)
+
+        if not full_graph_.has_edge(X, Y):
+            raise ValueError(f"The edge from {X} -> {Y} does not exist in the graph")
+
+        if (X in exposures) and (Y in observed):
+            if full_graph_.has_edge(X, Y):
+                full_graph_.remove_edge(X, Y)
+            dependent_var = Y
+
+        elif Y in latent_variables:
+            full_graph_.add_edge(Y, scaling_indicators[Y])
+            dependent_var = scaling_indicators[Y]
+
+        else:
+            dependent_var = Y
+
+        variable_parents = [
+            var for var in causal_graph.predecessors(Y) if not var.startswith(".")
+        ]
+
+        for parent_y in variable_parents:
+            if full_graph_.has_edge(X, Y):
+                full_graph_.remove_edge(X, Y)
+            if parent_y in latent_variables:
+                full_graph_.add_edge(scaling_indicators[parent_y], dependent_var)
+
+        return full_graph_, dependent_var
+
+    def _identify(self, causal_graph):
+        exposure = causal_graph.get_role("exposures")[0]
+        outcome = causal_graph.get_role("outcomes")[0]
+        observed = causal_graph.get_role("observed")
```

Comment:
I don't think we currently set the observed role by default. With multi-role support, handling both observed and latent roles might be a bit complicated. So, it might be better that we simply assume variables to be observed if the latent role is not set.